src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -753,10 +759,29 @@ ExecuteSingleModifyTask(CitusScanState *scanState, Task *task, bool multipleTask
 
 	char *queryString = task->queryString;
 	bool taskRequiresTwoPhaseCommit = (task->replicationModel == REPLICATION_MODEL_2PC);
+	bool failOnError = false;
 
 	ShardInterval *shardInterval = LoadShardInterval(task->anchorShardId);
 	Oid relationId = shardInterval->relationId;
 
+	if (multipleTasks)
+	{
+		taskRequiresTwoPhaseCommit |= (MultiShardCommitProtocol == COMMIT_PROTOCOL_2PC);
+	}
+
+	/* we don't allow failures to happen with CREATE INDEX CONCURRENTLY */
+	if (MultiShardCommitProtocol == COMMIT_PROTOCOL_BARE)
+	{
+		failOnError = true;
+	}
+
+	if (scanState)
+	{
+		operation = scanState->distributedPlan->operation;
+		executorState = scanState->customScanState.ss.ps.state;
+		paramListInfo = executorState->es_param_list_info;
+	}
+
 	/*
 	 * Modifications for reference tables are always done using 2PC. First
 	 * ensure that distributed transaction is started. Then force the
```

Comment:
There are some confusing interactions between `multipleTasks`, `failOnError`, `taskRequiresTwoPhaseCommit`, and (deeper down) `markCritical`. This could use some clean-up.

I wonder whether we should move the block below to the relevant caller.

Whether we're using 2PC seems in no way relevant to the execution of a single task, so it's kind of surprising that we have a `taskRequiresTwoPhaseCommit` variable in here. The fact that we don't deactivate a shard placement when a write to a reference table fails (*fail on error*) also seems somewhat independent from the fact that we execute it in a 2PC.

Reply:
I was thinking something along the lines of:
https://github.com/citusdata/citus/commit/c135a291bb4cd3a17596d4eb4779879e3b899a43

This lays out the different failure handling cases in `RouterSequentialModifyExecScan`.